### PR TITLE
[yargs] Update yargs constructor

### PR DESCRIPTION
**Pull request by jly36963**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [see comments]
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/yargs/index.d.mts [hunk 1]
```diff
@@ -8,7 +8,11 @@ interface MainType {
   filename: string;
   children: MainType[];
 }
-declare const _instanceFactory: (processArgs: string[], cwd?: string, parentRequire?: RequireType) => yargs.Argv;
+declare const _instanceFactory: (
+  processArgs?: ReadonlyArray<string> | string,
```
Comment (jly36963) [inline types/yargs/index.d.mts:12]: Currently, ts returns an error if a string or no value is passed to the yargs constructor; however, doing so is valid when using yargs.

[yargs source code](https://github.com/yargs/yargs/blob/8912078ae5e2157f5b78298a2ceb5cdf41518e85/lib/yargs-factory.ts#L67)

## Examples

All three of these are valid ways to execute yargs

```ts
import yargs from 'yargs'

// yargs with string

yargs('add 1 2')
  .command(
    "add <a> <b>",
    "add desc",
    (yargs) =>
      yargs
        .positional("a", {
          desc: "a desc",
          type: "number",
        })
        .positional("b", {
          desc: "a desc",
          type: "number",
        }),
    (argv) => {
      const { a, b } = argv as { a: number; b: number };
      console.log(`${a} + ${b} = ${a + b}`);
    },
  ).parse()

// yargs with array

yargs(['add', '1', '2'])
  .command(
    "add <a> <b>",
    "add desc",
    (yargs) =>
      yargs
        .positional("a", {
          desc: "a desc",
          type: "number",
        })
        .positional("b", {
          desc: "a desc",
          type: "number",
        }),
    (argv) => {
      const { a, b } = argv as { a: number; b: number };
      console.log(`${a} + ${b} = ${a + b}`);
    },
  ).parse()

// parse with string (also works with array)

yargs()
  .command(
    "add <a> <b>",
    "add desc",
    (yargs) =>
      yargs
        .positional("a", {
          desc: "a desc",
          type: "number",
        })
        .positional("b", {
          desc: "a desc",
          type: "number",
        }),
    (argv) => {
      const { a, b } = argv as { a: number; b: number };
      console.log(`${a} + ${b} = ${a + b}`);
    },
  ).parse('add 1 2')
```

File: types/yargs/ts4.1/yargs-tests.ts [hunk 2]
```diff
@@ -123,8 +123,7 @@ async function Argv$parsing() {
     const argv1 = await yargs.parse();
     const argv2 = yargs(['-x', '1', '-y', '2']).parseSync();
     const argv3 = yargs.parseSync(['-x', '1', '-y', '2']);
-    const argv4 = await yargs();
```
Comment (jly36963) [inline types/yargs/ts4.1/yargs-tests.ts:126]: AFAIK, `yargs()` does not return a promise

Comment (jly36963) [reply types/yargs/ts4.1/yargs-tests.ts:126]: Also, if `argv4` were kept, it should be the following:
```js
const argv4 = yargs().argv
```

The result of`yargs()` does not return `argv` -- it has `argv` as a property

Comment (jly36963) [reply types/yargs/ts4.1/yargs-tests.ts:126]: [parseAsync](http://yargs.js.org/docs/#api-reference-parseasyncargs-context-parsecallback) makes sense to use the `await` keyword
```js
const argv4 = await yargs().parseAsync()
```

**Review by bcoe — APPROVED** [review]

File: types/yargs/index.d.ts [hunk 3]
```diff
@@ -44,8 +44,7 @@ declare namespace yargs {
      * `Arguments<T>` to simplify the inferred type signature in client code.
      */
     interface Argv<T = {}> {
-        (): { [key in keyof Arguments<T> as key | CamelCaseKey<key>]: Arguments<T>[key] } | Promise<{ [key in keyof Arguments<T> as key | CamelCaseKey<key>]: Arguments<T>[key] }>;
```
Comment (sandersn) [inline types/yargs/index.d.ts:47]: Deleting this line implies that using `Arguments<T>` is incorrect for the no-argument case. Is that correct? I don't know yargs well, but this seems like a big change.

**Review by sandersn — CHANGES_REQUESTED** [review]
One question, although this really needs a yargs expert to look at it.

Edit: Oh, didn't see that @bcoe already did.

File: types/yargs/index.d.ts (hunk 3 above)
Comment (jly36963) [reply types/yargs/index.d.ts:47]: Let me look a little bit further -- I need to make sure I chose the correct one between `Argv<T>` and `{ [key in keyof Arguments<T> as key | CamelCaseKey<key>]: Arguments<T>[key] }`

Things I know:
- the return type should not be a promise
- `yargs()`, `yargs('hello')`, `yargs(['hello'])` should all have the same generic return type. (I still need to determine which one of the above it is)

**Comment by jly36963** [comment]
I'm confident in the changes I made.

## Foreword

```ts
{ [key in keyof Arguments<T> as key | CamelCaseKey<key>]: Arguments<T>[key] } 
  | Promise<{ [key in keyof Arguments<T> as key 
  | CamelCaseKey<key>]: Arguments<T>[key] }>
```

The above type is really long , and I'm going to refer to it as DefinitelyTyped Arguments.

Also, it should be noted that yargs type Argv (parsed arguments) and DefinitelyTyped type Argv (the yargs instance) mean different things. This might make the below explanation confusing.


## yargsInstance vs argv

```js
import yargs from 'yargs'

const yargsInstance = yargs() // DefinitelyTyped Argv<T>
const argv = yargsInstance.argv // DefinitelyTyped Arguments
```

[DefinitelyTyped Argv](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/yargs/index.d.ts#L46) is the equivalent of [YargsInstance](https://github.com/yargs/yargs/blob/main/lib/yargs-factory.ts#L164). The constructor should return a yargs instance, not the parsed arguments.

[DefinitelyTyped Arguments](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/yargs/index.d.ts#L461) is the equivalent of [Arguments](https://github.com/yargs/yargs/blob/main/lib/yargs-factory.ts#L1076) 

## Return types

### Yargs instance

DefinitelyTyped Argv is the correct return type for the following:
- `yargs`
- `yargs()`
- `yargs('something')`
- `yargs(['something'])`

### Arguments

DefinitelyTyped Arguments is the correct return type for the following:
- `yargs.argv`
- `yargs.parse()`

(`parseSync` and `parseAsync` will return the type-narrowed version of the return type for `.parse()`)

**Comment by jly36963** [comment]
I updated the tests to reflect my changes. Let me know if I need to add additional tests. (I'm unfamiliar with how the tests in DefinitelyTyped are written, so I might need some advice or a link to something I can educate myself with.)

**Review by andrewbranch — APPROVED** [review]
Checked against runtime behavior and it looks right. It does seem like `Argv<T>` is a pretty bad name, and might be nice to update it to `YargsInstance` in the future.

**Comment by jly36963** [comment]
Ready to merge

**Pull request merged**